File: internal/models/metrics.go
package models

import "time"

// Pool status constants
const (
	StatusHealthy = "healthy"  // HikariCP metrics available
	StatusNoPool  = "no_pool"  // Service alive but no connection pool
	StatusError   = "error"    // Service unreachable or error
)

// PoolMetrics represents connection pool and JVM metrics at a point in time
type PoolMetrics struct {
	ID           int64     `json:"id"`
	TargetName   string    `json:"target_name"`
	InstanceName string    `json:"instance_name"`
	Status       string    `json:"status"` // healthy, no_pool, error

	// HikariCP metrics
	Active     int     `json:"active"`
	Idle       int     `json:"idle"`
	Pending    int     `json:"pending"`
	Max        int     `json:"max"`
	Timeout    int64   `json:"timeout"`
	AcquireP99 float64 `json:"acquire_p99"`

	// JVM metrics
	HeapUsed    int64   `json:"heap_used"`     // bytes
	HeapMax     int64   `json:"heap_max"`      // bytes
	NonHeapUsed int64   `json:"non_heap_used"` // bytes
	NonHeapMax  int64   `json:"non_heap_max"`  // bytes
	ThreadsLive int     `json:"threads_live"`
	CpuUsage    float64 `json:"cpu_usage"` // 0.0 ~ 1.0

	// GC metrics
	GcCount     int64   `json:"gc_count"`      // total GC count
	GcTime      float64 `json:"gc_time"`       // total GC time in seconds
	YoungGcCount int64  `json:"young_gc_count"` // young gen GC count
	OldGcCount   int64  `json:"old_gc_count"`   // old gen GC count

	Timestamp time.Time `json:"timestamp"`
}

// TargetStatus represents current status of a monitoring target
type TargetStatus struct {
	Name      string           `json:"name"`
	Group     string           `json:"group,omitempty"` // Environment group: dev, staging, prod, etc.
	Status    string           `json:"status"`          // healthy, unhealthy, unknown
	Current   *PoolMetrics     `json:"current,omitempty"`
	Instances []InstanceStatus `json:"instances,omitempty"`
}

// InstanceStatus represents current status of an instance
type InstanceStatus struct {
	InstanceName string       `json:"instance_name"`
	Status       string       `json:"status"`
	Current      *PoolMetrics `json:"current,omitempty"`
}

// HistoryResponse represents historical metrics data
type HistoryResponse struct {
	TargetName string        `json:"target_name"`
	Datapoints []PoolMetrics `json:"datapoints"`
}

File: web/src/types/metrics.ts
export interface PoolMetrics {
  id: number;
  target_name: string;
  instance_name: string;
  status: string;
  active: number;
  idle: number;
  pending: number;
  max: number;
  timeout: number;
  acquire_p99: number;
  // JVM metrics
  heap_used: number;
  heap_max: number;
  non_heap_used: number;
  threads_live: number;
  cpu_usage: number;
  // GC metrics
  gc_count: number;
  gc_time: number;
  young_gc_count: number;
  old_gc_count: number;
  timestamp: string;
}

export interface InstanceStatus {
  instance_name: string;
  status: 'healthy' | 'warning' | 'critical' | 'unknown';
  current?: PoolMetrics;
}

export interface TargetStatus {
  name: string;
  group?: string;
  status: 'healthy' | 'warning' | 'critical' | 'unknown';
  current?: PoolMetrics;
  instances?: InstanceStatus[];
}

export interface TargetsResponse {
  targets: TargetStatus[];
  groups?: string[];
}

export interface HistoryResponse {
  target_name: string;
  datapoints: PoolMetrics[];
}

export interface Recommendation {
  type: string;
  current: string;
  recommended: string;
  reason: string;
  severity: 'info' | 'warning' | 'critical';
}

export interface PoolStats {
  avg_active: number;
  max_active: number;
  avg_idle: number;
  avg_pending: number;
  max_pending: number;
  avg_usage: number;
  peak_usage: number;
  current_max: number;
  timeout_count: number;
}

export interface AnalysisResult {
  target_name: string;
  analyzed_at: string;
  data_points: number;
  recommendations: Recommendation[];
  stats: PoolStats;
}

export interface LeakAlert {
  type: string;
  severity: string;
  message: string;
  detected_at: string;
  duration: string;
  avg_active: number;
  avg_idle: number;
  suggestions: string[];
}

export interface LeakAnalysisResult {
  target_name: string;
  analyzed_at: string;
  data_points: number;
  has_leak: boolean;
  leak_risk: 'none' | 'low' | 'medium' | 'high' | 'unknown';
  alerts: LeakAlert[];
  health_score: number;
}

// Alert types
export interface Alert {
  id: number;
  target_name: string;
  instance_name: string;
  rule_name: string;
  severity: 'info' | 'warning' | 'critical';
  message: string;
  status: 'fired' | 'resolved';
  fired_at: string;
  resolved_at?: string;
  notified_at?: string;
  channels?: string;
}

export interface AlertsResponse {
  alerts: Alert[];
}

export interface AlertStats {
  total_alerts: number;
  active_alerts: number;
  resolved_alerts: number;
  critical_count: number;
  warning_count: number;
  info_count: number;
}

// Alert Rule types
export interface AlertRule {
  id: number;
  name: string;
  condition: string;
  severity: 'info' | 'warning' | 'critical';
  message: string;
  enabled: boolean;
  created_at: string;
  updated_at: string;
}

export interface AlertRuleInput {
  name: string;
  condition: string;
  severity: 'info' | 'warning' | 'critical';
  message: string;
  enabled?: boolean;
}

export interface AlertRulesResponse {
  rules: AlertRule[];
  config_rules: {
    name: string;
    condition: string;
    severity: string;
    message: string;
    enabled?: boolean;
  }[];
}
